flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnalignerTest.java

```diff
@@ -639,4 +673,79 @@ public long getLastCanceledCheckpointId() {
 			return lastCanceledCheckpointId;
 		}
 	}
+
+	/**
+	 * Specific {@link AbstractInvokable} implementation to record and validate which checkpoint
+	 * id is executed and how many checkpoints are executed.
+	 */
+	private static final class ValidatingCheckpointInvokable extends AbstractInvokable {
+
+		private long expectedCheckpointId;
+
+		private int totalNumCheckpoints;
+
+		ValidatingCheckpointInvokable() {
+			super(new DummyEnvironment("test", 1, 0));
+		}
+
+		public void invoke() {
+			throw new UnsupportedOperationException();
+		}
+
+		public void triggerCheckpointOnBarrier(
+				CheckpointMetaData checkpointMetaData,
+				CheckpointOptions checkpointOptions,
+				CheckpointMetrics checkpointMetrics) {
+			expectedCheckpointId = checkpointMetaData.getCheckpointId();
+			totalNumCheckpoints++;
+		}
+
+		@Override
+		public <E extends Exception> void executeInTaskThread(
+				ThrowingRunnable<E> runnable,
+				String descriptionFormat,
+				Object... descriptionArgs) throws E {
+			runnable.run();
+		}
+
+		long getTriggeredCheckpointId() {
+			return expectedCheckpointId;
+		}
+
+		int getTotalTriggeredCheckpoints() {
+			return totalNumCheckpoints;
+		}
+	}
+
+	/**
+	 * Specific {@link CheckpointBarrierUnaligner} implementation to mock the scenario that the later triggered
+	 * checkpoint executes before the preceding triggered checkpoint.
+	 */
+	private static final class ValidatingCheckpointBarrierUnaligner extends CheckpointBarrierUnaligner {
+
+		private ThrowingRunnable waitingRunnable;
+		private boolean firstRunnable = true;
+
+		ValidatingCheckpointBarrierUnaligner(AbstractInvokable invokable) {
+			super(
+				new int[]{1},
+				new ChannelStateWriter.NoOpChannelStateWriter(),
+				"test",
+				invokable);
+		}
+
+		@Override
+		protected <E extends Exception> void executeInTaskThread(
+				ThrowingRunnable<E> runnable,
+				String descriptionFormat,
+				Object... descriptionArgs) throws E {
+			if (firstRunnable) {
+				waitingRunnable = runnable;
+				firstRunnable = false;
+			} else {
+				super.executeInTaskThread(runnable, "checkpoint");
+				super.executeInTaskThread(waitingRunnable, "checkpoint");
+			}
+		}
+	}
```

Comment:
I think this is not the best way to test this race condition. It's dubious to override a class that we want to test here (`ValidatingCheckpointBarrierUnaligner`). Also this is a very tight coupling and depending on the private implementation details, which assumes `executeInTaskThread` will be called only once. More over it brakes an assumption that mails should be executed in order.

I think it would be much better to do in the following manner:

1. straighten up threading model a bit and do not enqueue any emails in `CheckpointBarrierUnaligner#processBarrier` call, as this is already happening inside mailbox thread. It introduces unnecessary possibility for the race conditions and makes our live so much more difficult in this test. Currently by going through mailbox in `processBarrier` we are avoiding a bit of code duplication/simplifing a bit `notifyBarrierReceived` method, in an exchange of bad threading model and actually IMO more complicated code. We should fix this either way.

2. pass an `AbstractInvokable` instance that would be using `SteppingMailboxProcessor` to implement `org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable#executeInTaskThread` 

3. inside the `testConcurrentProcessBarrierAndNotifyBarrierReceived` test do the following sequence:
```
handler.getThreadSafeUnaligner().notifyBarrierReceived(buildCheckpointBarrier(0), channelInfo); // (a)
handler.processBarrier(buildCheckpointBarrier(1), 0); // (b)
steppingMailboxExecutor.runMailboxStep(); // (c)
```
(c) would execute mailbox action from (a). 

This would test the race condition without braking any contracts (like out of order mail execution) and without overriding `CheckpointBarrierUnaligner`.

Reply:
I indeed considered the way of verifying the race condition via somehow real `AbstractInvokable` with `TaskMailbox`, but also thought that these two components are a bit far away from `CheckpointBarrierHandler` and they are also a bit heavy-weight components from themselves. 

From the aspect of touching less external components in unit tests, i chose the current way. Actually I bypassed the mailbox implementation and simulate the race condition via executing the runnable in mis-order way. The propose for introducing `ValidatingCheckpointInvokable` and `ValidatingCheckpointBarrierUnaligner` is just for avoiding relying on external components of `AbstractInvokable` and `TaskMailbox` in unit tests.

And this test is for verifying the processes of `CheckpointBarrierUnaligner#processBarrier` and `#notifyBarrierReceived`, to confirm the new introduced method `CheckpointBarrierUnaligner#notifyCheckpoint` really effect in these interactions. All these three methods would be really touched in this test.

From another aspect, for the interaction between two components it is better to verify the real interactions using two real components without re-implementing either sides. Then any internal core changes in either component will be reflected in the test. For this case, actually the `CheckpointBarrierUnaligner` component will interact with `AbstractInvokable` with internal `TaskMailbox` model.  `SteppingMailboxProcessor` is also a re-implemented model to replace the real component inside `AbstractInvokable`, so it somehow still relies on the private implementation inside `SteppingMailboxProcessor`.

All in all, it might be better than my current way, and i can try out to use the real model AMAP in this test.

Reply:
I wasn't meaning to pass a real `AbstractInvocable`, but an equivalent of `DummyInvokable` that has a `SteppingMailboxProcessor`

> For this case, actually the CheckpointBarrierUnaligner component will interact with AbstractInvokable with internal TaskMailbox model. SteppingMailboxProcessor is also a re-implemented model to replace the real component inside AbstractInvokable, so it somehow still relies on the private implementation inside SteppingMailboxProcessor.

It's not the same. 

1. In your case, you are modifying the class that you are testing, which invalidates the test to some extent.

2. The idea of mocking (for example passing `DummyInvokable` with `SteppingMailboxProcessor`) is that you provide a mock implementation of some interface, that adheres to the contract of those interfaces (maybe in limited/restricted scope, but it shouldn't be braking them) and you pass them to real production code that you intend to test.

For example imagine a simple change, if `CheckpointBarrierUnaligner` changes and enqueues one more mail in one of the calls. Or that it starts relying on the fact, that order of enqueued emails guarantees their execution order - both of those refactorings would be valid but wouldn't work with your current implementation.